src/fancyflow.erl

```diff
@@ -1,29 +1,184 @@
+
 %% API exports
--export([pipe/2, maybe/2, parallel/1]).
+-export([parse_transform/2]).
+-export([maybe/2, parallel/1]).
+
+-type op_kind() :: mapper | folder.
+-record(op, {kind :: op_kind()
+            ,mname :: module()
+            ,fname :: atom()
+            }).
+
+-ifdef(fun_stacktrace).
+-define(WITH_STACKTRACE(E,T,ST), E:T -> ST = erlang:get_stacktrace(),).
+-else.
+-define(WITH_STACKTRACE(E,T,ST), E:T:ST ->).
+-endif.
 
 %%====================================================================
-%% API functions
+%% Parse Transform
+%%====================================================================
+
+parse_transform(ASTs, _Options) ->
+    try [erl_syntax_lib:map(fun revert_then_transform/1, AST) || AST <- ASTs]
+    catch _E:_R -> ASTs
+    end.
+
+revert_then_transform(T) ->
+    transform(erl_syntax:revert(T)).
+
+%% [pipe](Var, F1, F2)
+%% [parallel](A,B,C,D)
+%% [{folder,knm_numbers,pipe}](T,F1,F2,F3)
+transform({call, Line,
+           {cons, CLine, Operator={_,CLine,_}, {nil,CLine}},
+           InitAndFunsOrJustFuns}) when is_list(InitAndFunsOrJustFuns) ->
+    Op = op_new(Operator),
+    op_mixin(Op, InitAndFunsOrJustFuns, Line, CLine);
+transform(Term) ->
+    Term.
+
+op_new({atom,_,pipe}) -> #op{kind=folder, mname=?MODULE, fname=pipe};
+op_new({atom,_,maybe}) -> #op{kind=folder, mname=?MODULE, fname=maybe};
+op_new({atom,_,parallel}) -> #op{kind=mapper, mname=?MODULE, fname=parallel};
+op_new({tuple, _, [{atom,_,K}, {atom,_,M}, {atom,_,F}]}) ->
+    #op{kind=K, mname=M, fname=F}.
+
+op_mixin(#op{mname=?MODULE, fname=pipe}, InitAndFuns, Line, _) ->
+    mixin_pipe(InitAndFuns, Line);
+op_mixin(#op{mname=?MODULE, fname=maybe}, InitAndFuns, Line, _) ->
+    mixin_maybe(InitAndFuns, Line);
+%%TODO: inline [parallel]
+op_mixin(#op{kind=folder, mname=M, fname=F}, [Init|Funs=[_|_]], Line, CLine) ->
+    NewFuns = folder_funs(Funs, CLine),
+    Operation = {remote, Line, {atom,CLine,M}, {atom,CLine,F}},
+    {call, Line, Operation, [Init,NewFuns]};
+op_mixin(#op{kind=mapper, mname=M, fname=F}, Funs=[_|_], Line, CLine) ->
+    NewFuns = mapper_funs(Funs, CLine),
+    Operation = {remote, Line, {atom,CLine,M}, {atom,CLine,F}},
+    {call, Line, Operation, [NewFuns]}.
+
+%% @doc
+%% Inlines fancyflows' "pipe" which is defined as:
+%% -spec pipe(State, [fun((State) -> State)]) -> State when
+%%       State :: any().
+%% pipe(Init, []) -> Init;
+%% pipe(Init, [_|_]=Funs) ->
+%%     Apply = fun(F, State) -> F(State) end,
+%%     lists:foldl(Apply, Init, Funs).
+%% @end
+mixin_pipe([Init], _) -> Init;
+mixin_pipe([Init|Funs=[_|_]], Line) ->
+    Var0 = {var, Line, make_var_name()},
+    Expr0 = {match, Line, Var0, Init},
+    Acc0 = {Var0, [Expr0]},
+    {LastVar,Exprs} = lists:foldl(fun mixin_pipe_fold/2, Acc0, Funs),
+    Block = lists:reverse(Exprs, [LastVar]),
+    {block, Line, Block}.
+
+mixin_pipe_fold(Piped, {{var,_,LastVarName},Block}) ->
+    L = element(2, Piped),
+    Replacer = fun (T) -> replace_var(T, LastVarName) end,
```

Comment:
I think this form of rewriting may break the cases where the var refers to an external scope thing vs. introducing a new one. In:
```
A = 5,
[pipe](Val, begin X = A + _, X*2 end, begin X = A - _, X*3 end)
``` 
we'd expect to rewrite `X` to avoid exporting what is currently scoped to a function, but not `A`, which comes from the parent scope.

I can't figure out if the error is there, but the lack of scope analysis in the inlining of the macro makes me think it's not there.

Reply:
Unless I'm getting it wrong and what this does is inline the call that the fancyflow module does, and all individual clauses are still walled within funs

Reply:
Thanks for reviewing!
You are right the inlining does mess up the internal scope (I'm adding your example as a test).
However, `[{K,M,F}]` is not inlined (mixin-ed) and should be enclosing each expr in a fun. Turns out that also fails the new tests.

I'm looking into this.